// slimcore-3.0/lib/slimcore.js
/// <reference path="slimcore.d.ts" />
"use strict";
const version = require("./version");
const events_1 = require("events");
const native = module.require('../bin/slimcore.node');
function isMatch(value, pattern) {
    // tslint:disable-next-line:no-for-in
    for (const key in pattern) {
        if (!pattern.hasOwnProperty(key)) {
            continue;
        }
        if (value[key] !== pattern[key]) {
            return false;
        }
    }
    return true;
}
function makeFilteredListener(listener, filter) {
    return (args) => {
        if (isMatch(args, filter)) {
            listener(args);
        }
    };
}
class EventEmitterEx extends events_1.EventEmitter {
    handle(event, filter, listener) {
        const callback = filter ? makeFilteredListener(listener, filter) : listener;
        this.addListener(event, callback);
        return {
            dispose: () => {
                this.removeListener(event, callback);
            },
        };
    }
}
// version
exports.getVersion = version.getVersion;
exports.getApiVersion = version.getApiVersion;
// slimcore
exports.createSlimCoreInstance = native.createSlimCoreInstance;
exports.createChromiumFrameSink = native.createChromiumFrameSink;
exports.createCanvasFrameSink = native.createCanvasFrameSink;
exports.getInstanceCounts = native.getInstanceCounts;
exports.queryDeviceRotation = native.queryDeviceRotation;
exports.getAuthorizationStatus = native.getAuthorizationStatus;
exports.requestAuthorization = native.requestAuthorization;
const classes = [
    native.SlimCore,
    native.CallHandler,
    native.VideoBinding,
    native.VideoBindingRenderer,
    native.VideoBindingScreenShare,
    native.FrameSink,
    native.CanvasFrameSink,
    native.ChromiumFrameSink,
    native.ContentSharing,
    native.DataChannel,
    native.DataSource,
    native.DataSink,
    native.Trouter,
    native.TrouterListener,
    native.TrouterRequest,
    native.TrouterResponse,
    native.Ndi,
    native.NdiFrameSink,
    native.NdiParticipantStream,
];
for (const cls of classes) {
    if (Object.getPrototypeOf(cls.prototype) === events_1.EventEmitter.prototype) {
        Object.setPrototypeOf(cls.prototype, EventEmitterEx.prototype);
    }
    exports[cls.name] = cls;
}
exports.Enums = native.Enums;
// CallHandler.prototype.getProperties
function invoke(expr, fallback) {
    try {
        return expr();
    }
    catch (error) {
        return fallback;
    }
}
// tslint:disable-next-line:max-line-length
function getProperties(callHandler, objectId, strProperties, intProperties) {
    const result = {};
    markObjectSimple(result);
    for (const [key, item] of Object.entries(strProperties)) {
        result[key] = invoke(() => callHandler.getStrProperty(item.objectId || objectId, item.propKey), item.fallback);
    }
    for (const [key, item] of Object.entries(intProperties)) {
        result[key] = invoke(() => callHandler.getIntProperty(item.objectId || objectId, item.propKey), item.fallback);
    }
    return result;
}
// tslint:disable-next-line:max-line-length
native.CallHandler.prototype.getProperties = function (objectId, strProperties, intProperties) {
    // tslint:disable-next-line:no-invalid-this
    return getProperties(this, objectId, strProperties, intProperties);
};
function markObjectSimple(object) {
    native._setHiddenProperty(object, 'simple', true);
}
function setReturnType(object, type) {
    native._setHiddenProperty(object, 'returnType', type);
}
// constant
setReturnType(exports.getVersion, 'constant');
setReturnType(exports.getApiVersion, 'constant');
setReturnType(native.SlimCore.prototype.getBuildName, 'constant');
setReturnType(native.SlimCore.prototype.getBuildVersion, 'constant');
// promise
setReturnType(native.SlimCore.prototype.getMonitorSnapshot, 'promise');
setReturnType(native.SlimCore.prototype.getWindowSnapshot, 'promise');
setReturnType(native.SlimCore.prototype.getWindowIcon, 'promise');
setReturnType(native.SlimCore.prototype.videoCreateBinding, 'promise');
setReturnType(native.SlimCore.prototype.videoReleaseBinding, 'promise');
setReturnType(native.VideoBindingRenderer.prototype.captureFrame, 'promise');
// void
setReturnType(native.FrameSink.prototype.log, 'void');
setReturnType(native.SlimCore.prototype.flushLogs, 'void');
// backwards compatibility
native.Ndi.prototype.observeLogIn = () => { };
native.Ndi.prototype.observeLogOut = () => { };
native.SlimCore.prototype.hasFeature = (feature) => {
    return feature in SlimCore.Enums.Feature;
};
